### PR TITLE
Cleanup/improve C style memory management tests

### PR DESCRIPTION
**Pull request by dalg24**
Tidying things up in the process of revisiting all *_NO_THROW assertions

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
(clang-format removed trailing empty line)

File: core/unit_test/Makefile [hunk 1]
```diff
@@ -392,7 +392,6 @@ ifneq ($(KOKKOS_INTERNAL_COMPILER_HCC), 1)
   OBJ_DEFAULT += TestDefaultDeviceType_a1.o TestDefaultDeviceType_b1.o TestDefaultDeviceType_c1.o
   OBJ_DEFAULT += TestDefaultDeviceType_a2.o TestDefaultDeviceType_b2.o TestDefaultDeviceType_c2.o
   OBJ_DEFAULT += TestDefaultDeviceType_a3.o TestDefaultDeviceType_b3.o TestDefaultDeviceType_c3.o
-  OBJ_DEFAULT += TestDefaultDeviceType_d.o
```
Comment (romintomasetti) [inline core/unit_test/Makefile:395]: Would that mean that if someone uses `makefile` only, this test disappears?

Comment (romintomasetti) [reply core/unit_test/Makefile:395]: Or maybe it's automatically included, haven't checked, just wondering.

File: core/unit_test/TestCStyleMemoryManagement.cpp [hunk 2]
```diff
@@ -14,28 +14,29 @@
...
 #include <TestDefaultDeviceType_Category.hpp>
-#include <TestUtilities.hpp>
 
-namespace Test {
+#include <gtest/gtest.h>
+
+namespace {
 
-TEST(defaultdevicetype, malloc) {
+TEST(defaultdevicetype, c_style_memory_management_malloc_realloc_and_free) {
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/TestCStyleMemoryManagement.cpp:25]: Its still in the "DefaultDevice" executable, and there are some other test files not named "DefaultDevice" something that also go into that test executable.

File: core/unit_test/Makefile (hunk 1 above)
Comment (dalg24) [reply core/unit_test/Makefile:395]: It means we are not testing when using generated makefiles

File: core/unit_test/TestCStyleMemoryManagement.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/unit_test/TestCStyleMemoryManagement.cpp:25]: You mean you would prefer if I kept the test file in the default/  subdirectory?  I don't feel too strongly about that one, I only saw that all test with a meaningful name were not segregated and did the same thing.

**Review by Rombur — APPROVED** [review]

File: core/unit_test/TestCStyleMemoryManagement.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/TestCStyleMemoryManagement.cpp:25]: I am ok with what you did here.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply core/unit_test/TestCStyleMemoryManagement.cpp:25]: OK too.

File: core/unit_test/TestCStyleMemoryManagement.cpp [hunk 3]
```diff
@@ -14,28 +14,29 @@
...
+  ASSERT_NE(data, nullptr);
+
+  data = static_cast<int*>(Kokkos::kokkos_realloc(data, 120 * sizeof(int)));
+  ASSERT_NE(data, nullptr);
+
   Kokkos::kokkos_free(data);
+}
 
+TEST(defaultdevicetype, c_style_memory_management_malloc_zero_byte_and_free) {
   int* data2 = static_cast<int*>(Kokkos::kokkos_malloc(0));
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/unit_test/TestCStyleMemoryManagement.cpp:36]: Agreeing with you but it is not worth changing and waiting on the CI

**Comment by dalg24** [comment]
Ignoring NVHPC segfault in the OpenACC build.  Everything else passed.

**Pull request merged**